codecs/h264_packet_test.go

```diff
@@ -293,31 +293,47 @@ func TestH264Payloader_Payload_SPS_and_PPS_handling(t *testing.T) {
 	}
 }
 
-func TestH264Payloader_PayloadAndDetectDataIframeAfterSpsPps(t *testing.T) {
+// simulates a frame with metadata: sps, pps and sei followed by iframe
+func TestH264Payloader_PayloadAndDetectDataIframeAfterSpsPpsSei(t *testing.T) {
 	pck := H264Payloader{}
+	idx := -1
 
 	// When packetizing SPS and PPS are emitted with following NALU
-	res := pck.Payload(1500, []byte{0x07, 0x00, 0x01})
+	res, idx := pck.PayloadAndDetectData(1500, []byte{0x07, 0x00, 0x01})
 	if len(res) != 0 {
 		t.Fatal("Generated payload should be empty")
 	}
+	if idx != -1 {
+		t.Fatal("frame includes only sps, expecting index to be -1")
+	}
 
-	res = pck.Payload(1500, []byte{0x08, 0x02, 0x03})
+	res, idx = pck.PayloadAndDetectData(1500, []byte{0x08, 0x02, 0x03})
 	if len(res) != 0 {
 		t.Fatal("Generated payload should be empty")
 	}
+	if idx != -1 {
+		t.Fatal("frame includes only pps, expecting index to be -1")
+	}
 
-	idx := -1
-	res, idx = pck.PayloadAndDetectData(1500, []byte{0x05, 0x04, 0x05})
+	// this simulates a case where sei preceeds the iframe data
+	res, idx = pck.PayloadAndDetectData(1500, []byte{
+		0x0, 0x0, 0x0, 0x1, // NAL START
+		0x26,            // NAL header byte
+		0x5, 0xff, 0x18, // SEI header - payload type + payload size bytes
+		0x8f, 0x4a, 0xf5, 0x58, 0xe4, 0x21, 0x45, 0xc9, 0xb0, 0xa5, 0x47, 0x0d, 0x7a, 0x3a, 0x74, 0xf9, // subtitle uuid payload
+		// subtitle payload with some NAL starts code in it
+		0x8, 0x0, 0x0, 0x1, 0x7, 0x0, 0x0, 0x0, 0x1, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x7, 0x8,
+		0x0, 0x0, 0x0, 0x1, 0x05, 0x04, 0x05}) // iframe
```

Comment:
would have been nice to see additional data nal after iframe to see you return the first one (As the code doe)... but ok I guess... took it too far with the subtitle for such a test... You're not testing subtitle nal parsing, you could have just add dummy payload... But I can't really complain. (I gave you a gift and thought you'll just paste it into the test...)